plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchange.java

```diff
@@ -216,62 +218,70 @@ private ListenableFuture<List<ExchangeSourceHandle>> createExchangeSourceHandles
-                            .filter(file -> file.getFilePath().startsWith(committedAttemptId + PATH_SEPARATOR, attemptIdOffset) && file.getFilePath().endsWith(DATA_FILE_SUFFIX))
                             .collect(toImmutableList());
 
-                    ImmutableMultimap.Builder<Integer, FileStatus> result = ImmutableMultimap.builder();
-                    for (FileStatus partitionFile : partitionFiles) {
-                        Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getFilePath()).getName());
-                        checkState(matcher.matches(), "Unexpected partition file: %s", partitionFile);
-                        int partitionId = Integer.parseInt(matcher.group(1));
-                        result.put(partitionId, partitionFile);
+                    if (committedMarkerFilePaths.isEmpty()) {
+                        throw new IllegalStateException(format("No committed attempts found under sink output path %s", sinkOutputPath));
                     }
-                    return result.build();
+
+                    for (String committedMarkerFilePath : committedMarkerFilePaths) {
+                        // Committed marker file path format: {sinkOutputPath}/{attemptId}/committed
+                        String[] parts = committedMarkerFilePath.split(PATH_SEPARATOR);
+                        checkState(parts.length >= 3, "committedMarkerFilePath %s is malformed", committedMarkerFilePath);
+                        String stringCommittedAttemptId = parts[parts.length - 2];
+                        int attemptIdOffset = committedMarkerFilePath.length() - stringCommittedAttemptId.length()
+                                - PATH_SEPARATOR.length() - COMMITTED_MARKER_FILE_NAME.length();
+
+                        // Data output file path format: {sinkOutputPath}/{attemptId}/{sourcePartitionId}_{splitId}.data
+                        List<FileStatus> partitionFiles = sinkOutputFiles.stream()
+                                .filter(file -> file.getFilePath().startsWith(stringCommittedAttemptId + PATH_SEPARATOR, attemptIdOffset) && file.getFilePath().endsWith(DATA_FILE_SUFFIX))
+                                .collect(toImmutableList());
+
+                        ImmutableMultimap.Builder<Integer, SourceFile> result = ImmutableMultimap.builder();
+                        for (FileStatus partitionFile : partitionFiles) {
+                            Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getFilePath()).getName());
+                            checkState(matcher.matches(), "Unexpected partition file: %s", partitionFile);
+                            int partitionId = parseInt(matcher.group(1));
+                            result.put(partitionId, new SourceFile(partitionFile.getFilePath(), partitionFile.getFileSize(), partition, parseInt(stringCommittedAttemptId)));
+                        }
+                        return result.build();
```

Comment:
at this commit this code is weird. Why do we have outer loop as we are always exiting from it in frist iteration. I clearly miss sth

Reply:
Oh - it used to be `findFirst`. Maybe drop `for` in this commit and only add it in the next one?

Reply:
Good point, changed